### ISSUE TITLE
Tactics fail when term tries to pattern match a tuple

### ISSUE THREAD

**Issue by Chris-Hawblitzel** [FStarLang/FStar#1347]
The following code fails:

```
module M
open FStar.Tactics

let foo () =
//  assert (let (y, z) = (1, 2) in True) // SUCCEEDS
//  assert_by_tactic (let yz = (1, 2) in True) smt // SUCCEEDS
  assert_by_tactic (let (y, z) = (1, 2) in True) smt // FAILS
```

The error is `(Error) assertion failed (see also FStar\ulib\FStar.Tactics.Effect.fst(102,45-102,55))`.  The commented out lines in the code above succeed.

The following code also fails:

```
open FStar.Tactics

let t1 : tactic unit =
  dump "Before";;
  u <-- forall_intro;
  smt

let fyz (yz:(int * int)) : Type = let (y, z) = yz in True

assume val g0 (x:int) : Ghost (int * int)
  (requires True)
//  (ensures fyz) // SUCCEEDS
//  (ensures (fun (yz:(int * int)) -> let xx = yz in True)) // SUCCEEDS
  (ensures (fun (yz:(int * int)) -> let (y, z) = yz in True)) // FAILS

let foo (x:nat) =
(
  let yz:(int * int) = g0 x in
  ()
) <: (Lemma True) by t1
```

The error is `(Error) user tactic failed: intro: unification failed (see also FStar\ulib\prims.fst(165,8-165,47)`.  By contrast, the commented out versions in the code above succeed.

The failure happens during `forall_intro`.  The output is:
```
proof-state: State dump @ depth 1 (Before):
Position: <dummy>(0,0-0,0)
ACTIVE goals (1):
(x:Prims.nat) |- (*?u18*) _ x : Prims.squash (*could not prove post-condition*)
  forall (p:Prims.pure_post Prims.unit).
    (forall (x:Prims.unit). p x) ==>
    (forall (x:Prims.int * Prims.int).
        (let #uu___610, #uu___611, y, z = x in Prims.l_True) ==>
        x == M.g0 x ==>
        (forall (y:Prims.int * Prims.int). y == x ==> (forall (x:Prims.unit). p x)))
SMT goals (0):

proof-state: State dump @ depth 1 (at the time of failure):
Position: C:\b\fstar\a2\FStar\ulib\FStar.Tactics.Builtins.fst(201,89-201,102)
ACTIVE goals (1):
(x:Prims.nat) |- (*?u21*) _ x : x:Prims.pure_post Prims.unit ->
Prims.Tot
(Prims.squash ((forall (x:Prims.unit). x x) ==>
      (forall (x:Prims.int * Prims.int).
          (let #uu___610, #uu___611, y, z = x in Prims.l_True) ==>
          x == M.g0 x ==>
          (forall (y:Prims.int * Prims.int). y == x ==> (forall (x:Prims.unit). x x)))))
SMT goals (1):
(x:Prims.nat) |- (*?u116*) _ x : Prims.squash (forall (x:Prims.pure_post Prims.unit).
      {:pattern
       could not prove post-condition
      }
      (forall (x:Prims.unit). x x) ==>
      (forall (x:Prims.int * Prims.int) (k:Prims.pure_post Type0).
          {:pattern
           could not prove post-condition
          }
          (forall (x:Type0). {:pattern  Prims.guard_free (k x) } k x <==> Prims.l_True) ==>
          (forall (b:Prims.int) (b:Prims.int).
              x == FStar.Pervasives.Native.Mktuple2 b b ==>
              (forall (y:Type0). y == Prims.l_True ==> k y))))
```

**Comment by mtzguido** [comment]
I'm sorry I took so long on this one, Chris.

The first problem was only tangentially related to matching. It turns out that on that VC we apply the (poorly named) `ite_wp` combinator which changes `wp` to `fun p -> forall k. (forall x. k x <==> p x) ==> wp k`, which is a way to get an equivalent `wp` that doesn't duplicate the postcondition. The thing is that tactics are triggered by polarity, that is, ignored at negative positions and run at positive ones, but I forgot to account for the fact that for `<==>` both of its arguments are bipolar. I changed the traversal to take this into account and your first example now passes.

It might help to know the way it works now. For a `p <==> q` subformula, the engine will descend on both arguments with both polarities (but in a single pass for each to avoid an exponential blowup). Then we get positively- and negatively-traversed versions of each, and turn the formula into `(p- ==> q+) /\ (q- ==> p+)`. That is, unless neither side had any tactics, in which case the formula is unchanged. Concretely, if you had
```
by_tactic t1 (squash phi) <==> xi /\ by_tactic t2 (squash psi)
```
the VC would turn to
```
    (phi ==> xi /\ True)
/\  (xi /\ psi ==> True)
```
plus the result of running both tactics.

The second error is because `match`es don't unify with themselves (or with anything I believe). I've ran into this a few times, but I'm not sure on how that can be fixed. Sadly this implies that if your goal contains a match, pretty much any primitive to change it would fail... I'll try to mention this in tomorrow's meeting to see what we can do.
